File: dotnet/DesignPatterns/CreationalDesignPatterns/PrototypeDesignPattern/PrototypeDesignPattern.Example1/Models/Concrete/ReportDocument.cs
using PrototypeDesignPattern.Example1.Models.Abstract;

namespace PrototypeDesignPattern.Example1.Models.Concrete;

public class ReportDocument : DocumentBase
{
    public ReportHeader Header { get; set; } = new();
    public List<ReportSection> Sections { get; set; } = new();
    public ReportSummary Summary { get; set; } = new();
    public ReportType ReportType { get; set; }

    public ReportDocument()
    {
        Title = "Report Template";
        Content = "Standard Report Template";
        Metadata.Tags.AddRange(new[] { "Report", "Analysis", "Corporate" });
    }

    public override bool ValidateDocument() =>
        !string.IsNullOrEmpty(Header.Title) &&
        !string.IsNullOrEmpty(Header.PreparedBy) &&
        Sections.Count > 0;

    public override string GetDocumentInfo() =>
        $"{base.GetDocumentInfo()}, Report Type: {ReportType}, Sections: {Sections.Count}, Prepared By: {Header.PreparedBy}";
}

public class ReportHeader
{
    public string Title { get; set; } = string.Empty;
    public string PreparedBy { get; set; } = string.Empty;
    public DateTime PreparedDate { get; set; } = DateTime.Now;
    public string Department { get; set; } = string.Empty;
}

public class ReportSection
{
    public string SectionTitle { get; set; } = string.Empty;
    public string Content { get; set; } = string.Empty;
    public List<string> KeyPoints { get; set; } = new();
    public int Order { get; set; }
}

public class ReportSummary
{
    public string ExecutiveSummary { get; set; } = string.Empty;
    public List<string> Recommendations { get; set; } = new();
    public List<string> Conclusions { get; set; } = new();
}

public enum ReportType
{
    Financial,
    Marketing,
    Technical,
    Management,
    Compliance
}

File: nodejs/DesignPatterns/CreationalDesignPatterns/PrototypeDesignPattern/PrototypeDesignPattern.Example1/src/models/concrete/ReportDocument.ts
import { DocumentBase } from '../abstract/DocumentBase';

export enum ReportType {
  Financial = 'Financial',
  Marketing = 'Marketing',
  Technical = 'Technical',
  Management = 'Management',
  Compliance = 'Compliance',
}

export class ReportHeader {
  title: string = '';
  preparedBy: string = '';
  preparedDate: Date = new Date();
  department: string = '';
}

export class ReportSection {
  sectionTitle: string = '';
  content: string = '';
  keyPoints: string[] = [];
  order: number = 0;
}

export class ReportSummary {
  executiveSummary: string = '';
  recommendations: string[] = [];
  conclusions: string[] = [];
}

export class ReportDocument extends DocumentBase {
  header: ReportHeader = new ReportHeader();
  sections: ReportSection[] = [];
  summary: ReportSummary = new ReportSummary();
  reportType: ReportType = ReportType.Management;

  constructor() {
    super();
    this.title = 'Report Template';
    this.content = 'Standard Report Template';
    this.metadata.tags.push('Report', 'Analysis', 'Corporate');
  }

  validateDocument(): boolean {
    return !!this.header.title && !!this.header.preparedBy && this.sections.length > 0;
  }

  override getDocumentInfo(): string {
    return `${super.getDocumentInfo()}, Report Type: ${this.reportType}, Sections: ${this.sections.length}, Prepared By: ${this.header.preparedBy}`;
  }
}
